pytest-embedded-idf/pytest_embedded_idf/dut.py

```diff
@@ -35,3 +48,31 @@ def toolchain_prefix(self) -> str:
             return f'riscv32-{self.target}-elf-'
         else:
             raise ValueError(f'Unknown target: {self.target}')
+
+    def image_info(self, path_to_file):
+        """
+        get hash256 of app
+
+        :param: path: path to file
+        :return: sha256 appended to app
+        """
+
+        old_stdout = sys.stdout
+        new_stdout = io.StringIO()
+        sys.stdout = new_stdout
+
+        class Args(object):
+            def __init__(self, attributes):
+                for key, value in attributes.items():
+                    self.__setattr__(key, value)
+
+        args = Args(
+            {
+                'chip': self.target,
+                'filename': path_to_file,
+            }
+        )
+        esptool.image_info(args)
+        output = new_stdout.getvalue()
```

Comment:
Does this return only the sha256 hash appended to the app? When I run `esptool.py image_info`, i get multiple lines of output instead:
```
esptool.py v3.3-dev
Image version: 1
Entry point: 400810d4
6 segments

Segment 1: len 0x07764 load 0x3f400020 file_offs 0x00000018 [DROM]
Segment 2: len 0x02674 load 0x3ffb0000 file_offs 0x00007784 [BYTE_ACCESSIBLE,DRAM]
Segment 3: len 0x06210 load 0x40080000 file_offs 0x00009e00 [IRAM]
Segment 4: len 0x146b0 load 0x400d0020 file_offs 0x00010018 [IROM]
Segment 5: len 0x05600 load 0x40086210 file_offs 0x000246d0 [IRAM]
Segment 6: len 0x00010 load 0x50000000 file_offs 0x00029cd8 [RTC_DATA]
Checksum: f4 (valid)
Validation Hash: d0cd46e5d6ba47189cb1360468d5e863fee9b91e52dff33d89a1f5d7b72fe135 (valid)
```

Looking at esptool code, we could possibly get just the SHA256 hash appended to the file using something like
```python
image = esptool.LoadFirmwareImage(self.target, path_to_file)
if image.append_digest:
    return image.stored_digest
```

However I don't know if we can rely on these esptool internals to be stable. OTOH, I am not sure we can rely on image_info output format to be stable, either. @radimkarnis what would you suggest?

Reply:
> Does this return only the sha256 hash appended to the app? When I run `esptool.py image_info`, i get multiple lines of output instead:
> 
> ```
> esptool.py v3.3-dev
> Image version: 1
> Entry point: 400810d4
> 6 segments
> 
> Segment 1: len 0x07764 load 0x3f400020 file_offs 0x00000018 [DROM]
> Segment 2: len 0x02674 load 0x3ffb0000 file_offs 0x00007784 [BYTE_ACCESSIBLE,DRAM]
> Segment 3: len 0x06210 load 0x40080000 file_offs 0x00009e00 [IRAM]
> Segment 4: len 0x146b0 load 0x400d0020 file_offs 0x00010018 [IROM]
> Segment 5: len 0x05600 load 0x40086210 file_offs 0x000246d0 [IRAM]
> Segment 6: len 0x00010 load 0x50000000 file_offs 0x00029cd8 [RTC_DATA]
> Checksum: f4 (valid)
> Validation Hash: d0cd46e5d6ba47189cb1360468d5e863fee9b91e52dff33d89a1f5d7b72fe135 (valid)
> ```
This output is correct. Just saw the `simpla_ota_example_test` and found that sha256 is searched in this output.


> However I don't know if we can rely on these esptool internals to be stable. OTOH, I am not sure we can rely on image_info output format to be stable, either. @radimkarnis what would you suggest?

This function is the same as in `ttfw_idf`.

Reply:
The thing is, ttfw_idf is kept in the IDF repository, where esptool dependency is expressed using a submodule. If esptool API changes, we will simply update ttfw_idf in the MR where esptool is updated.

However pytest_embedded version is not linked to a specific esptool version, which means that it should (at least try to) be forward and backward compatible. Let's see what esptool folks say.

Reply:
> This output is correct. Just saw the simpla_ota_example_test and found that sha256 is searched in this output.

Does it make sense to either rename the function and update the function description here (currently it claims to return just the sha256) or to make it return just the sha256, rather than the whole image_info output?

Reply:
This approach LGTM! the API will soon change to:
```python
image = esptool.bin_image.LoadFirmwareImage(self.target, path_to_file)
```
Other than that, the image_info output can be considered stable, there are no plans to change it.

Reply:
if the test requires to calculate the sha256 to a bin file, I think it's better to do it on the test script side, not adding a function here. the esptool call is not needed.

Reply:
@hfudev it's not exactly the sha256 of the bin file. It's SHA256 of most of the bin file, except for the SHA256 itself, which is appended at the end of the bin file :)

https://docs.espressif.com/projects/esptool/en/latest/esp32/advanced-topics/firmware-image-format.html#footer

(it is also optional, so some bin files might not have this hash appended)

Reply:
Now I get it. Thanks!